### PR TITLE
Synthesize pinch/rotation gestures from raw multi-touch events

### PR DESCRIPTION
**Pull request by tilladam**
## Summary

- Synthesize `PinchGesture`, `RotationGesture`, and `DoubleTapGesture` events from raw multi-touch input on backends that lack native gesture recognition (Android, Linux KMS)
- Add a `TouchState` state machine in `window.rs` that tracks active touches and transitions through Idle → TwoFingersDown → Pinching, producing the same `MouseEvent` variants that the winit backend already emits for macOS platform gestures
- Update the `native-gestures` example with Android support and improved UI
- Review fixes: sign convention normalization, fixed-capacity data structures (no heap allocation), callback ordering, deduplication of coordinate conversion and phase mapping helpers, edge-case test coverage

## Details

The `TouchState` recognizer:
- Forwards single-finger touches as mouse press/move/release events
- Detects two-finger gestures with configurable thresholds (8px pinch, 5° rotation)
- Synthesizes a release for the primary finger before entering gesture mode
- Re-presses the remaining finger when a gesture ends with one finger still down
- Detects double-tap with distance and timing constraints
- Uses fixed-capacity data structures (`TouchMap`, `TouchEventBuffer`) to avoid heap allocation

Backend changes:
- **Android**: Unified coordinate conversion, calls `process_touch_input` for multi-pointer events
- **Linux KMS**: Replaced `BTreeMap` with fixed-capacity array for touch positions
- **Winit**: Extracted `winit_touch_phase` helper, negates rotation delta at the source to match Slint convention (positive = clockwise)

## Test plan

- [x] `cargo test -p i-slint-core -- touch` (27 unit tests for TouchState)
- [x] `cargo test -p test-driver-rust -- pinchgesturehandler` (15 element tests)
- [x] Manual test: macOS trackpad pinch/rotation via `native-gestures` example
- [ ] Manual test: Android touchscreen via `native-gestures` example
- [ ] Manual test: Linux touchscreen

### CODE REVIEW

File: internal/backends/android-activity/androidwindowadapter.rs [hunk 1]
```diff
@@ -594,11 +628,28 @@ fn long_press_timeout() {
...
+/// display offset and scale factor.
+fn pointer_logical_position(
+    x: f32,
+    y: f32,
+    offset: PhysicalPosition,
+    scale_factor: f32,
+) -> LogicalPosition {
+    let phys_x = x - offset.x as f32;
+    let phys_y = y - offset.y as f32;
+    LogicalPosition::new(phys_x / scale_factor, phys_y / scale_factor)
```
Comment (tronical) [inline internal/backends/android-activity/androidwindowadapter.rs]: I think it would be cleaner here to retain the previous pattern: Create `PhysicalPosition` and use `to_logical`.

File: internal/backends/linuxkms/calloop_backend/input.rs [hunk 2]
```diff
@@ -117,7 +116,9 @@ pub struct LibInputHandler<'a> {
     libinput: input::Libinput,
     token: Option<calloop::Token>,
     mouse_pos: Pin<Rc<Property<Option<LogicalPosition>>>>,
-    last_touch_pos: LogicalPosition,
+    /// Last known position per touch slot. Fixed-capacity to avoid heap
```
Comment (tronical) [inline internal/backends/linuxkms/calloop_backend/input.rs]: I suggest to add a comment here explaining why this is needed. AFAICS it's because when we receive a touch up event, we don't know the position just from the event itself. (this isn't new in this PR, but would be great if you could do it while you're at it :)

File: internal/backends/linuxkms/calloop_backend/input.rs [hunk 3]
```diff
@@ -155,6 +156,29 @@ impl<'a> LibInputHandler<'a> {
     }
 }
 
+fn set_touch_pos(
```
Comment (tronical) [inline internal/backends/linuxkms/calloop_backend/input.rs:162]: Perhaps these two helper functions belong into LibInputHandler, then you don't need to repeat the `; 5]` size but just pass `&mut self`.

File: internal/core/window.rs [hunk 4]
```diff
@@ -438,13 +438,1023 @@ struct WindowPinnedFields {
...
+/// Fixed-capacity map of touch IDs to touch points.
+///
+/// Touchscreens rarely report more than 5 simultaneous contacts, and gesture
+/// recognition only uses the first two. A linear-scan array avoids the heap
+/// allocation and pointer-chasing overhead of `BTreeMap` for this tiny collection.
+const MAX_TRACKED_TOUCHES: usize = 5;
+
+#[derive(Clone)]
+struct TouchMap {
+    entries: [(u64, TouchPoint); MAX_TRACKED_TOUCHES],
```
Comment (tronical) [inline internal/core/window.rs]: An unnamed tuple that has a second field that's a named structure that in turn has just one field seems ... obscure to me :-)

Wouldn't it be more readable to have

```rust
struct TouchPoint {
    id: u64,
    position: LogicalPoint
}
```

and then an array of `[TouchPoint]` ?

File: internal/core/window.rs [hunk 5]
```diff
@@ -438,13 +438,1023 @@ struct WindowPinnedFields {
...
+        if self.len < MAX_TOUCH_EVENTS {
+            self.events[self.len] = Some(event);
+            self.len += 1;
+        }
+    }
+
+    #[cfg(test)]
+    fn iter(&self) -> impl Iterator<Item = &MouseEvent> {
+        self.events[..self.len].iter().filter_map(|e| e.as_ref())
+    }
```
Comment (tronical) [inline internal/core/window.rs]: Instead of this special casing, perhaps it would be simpler to make `TouchEventBuffer` Clone and let the test call `clone()` and then `into_iter()`.

File: internal/core/window.rs [hunk 6]
```diff
@@ -438,13 +438,1023 @@ struct WindowPinnedFields {
...
+/// When only one finger is down, touch events are forwarded as mouse events.
+/// When two fingers are down and move beyond a threshold, synthesized
+/// `PinchGesture` and `RotationGesture` events are emitted — the same events
+/// that platform gesture recognition (e.g. macOS trackpad) produces.
+struct TouchState {
+    active_touches: TouchMap,
+    /// The finger forwarded as mouse events during single-touch.
+    primary_touch_id: Option<u64>,
+    /// The two finger IDs participating in the gesture (when in TwoFingersDown or Pinching).
+    gesture_finger_ids: Option<(u64, u64)>,
```
Comment (tronical) [inline internal/core/window.rs]: I feel that it would be cleaner to have the two finger ids in the `GestureRecognitionState` enum itself, in the two variants where they're used. That way such a comment becomes unnecessary.

File: internal/core/window.rs [hunk 7]
```diff
@@ -438,13 +438,1023 @@ struct WindowPinnedFields {
...
+    gesture_state: GestureRecognitionState,
+    /// Distance between the two fingers when the gesture started.
+    initial_distance: f32,
```
Comment (tronical) [inline internal/core/window.rs]: I feel the same about these fields. Yes, they'd be duplicated within the variants, but it also makes it much safer to change them for just one variant without having to worry that they affect the state of the other.

File: internal/core/window.rs [hunk 8]
```diff
@@ -438,13 +438,1023 @@ struct WindowPinnedFields {
...
+
+    /// Minimum angular change (in degrees) before two fingers are recognized as a rotation.
+    const ROTATION_THRESHOLD: f32 = 5.0;
+
+    /// Maximum squared distance (in logical pixels) between two taps for a double-tap.
+    /// 100 px² ≈ 10px radius, matching [`ClickState::check_repeat`](crate::input::ClickState).
+    const DOUBLE_TAP_DISTANCE_SQ: f32 = 100.0;
+
+    /// Returns the positions of the two gesture fingers, or `None` if not available.
+    fn gesture_finger_positions(&self) -> Option<(&TouchPoint, &TouchPoint)> {
```
Comment (tronical) [inline internal/core/window.rs:611]: If the gesture_finger_ids were part of the enum, then this and the other helper functions would become more naturally functions of the state. (`match self` instead of `match self.gesture_state`). Not 1000% sure it's an improvement, but a guts feeling.

File: internal/core/window.rs [hunk 9]
```diff
@@ -438,13 +438,1023 @@ struct WindowPinnedFields {
...
+        let (id_a, id_b) = self.gesture_finger_ids?;
+        let a = self.active_touches.get(id_a)?;
+        let b = self.active_touches.get(id_b)?;
+        Some((a, b))
+    }
+
+    /// Returns the midpoint between the two gesture fingers, or `None`.
+    fn gesture_midpoint(&self) -> Option<LogicalPoint> {
+        let (a, b) = self.gesture_finger_positions()?;
+        Some(euclid::point2(
```
Comment (tronical) [inline internal/core/window.rs]: I think that you could perhaps write this as `self.gesture_finger_positions.and_then(|(a, b)| a.position.lerp(b.position, 0.5))` . 

More generally speaking here, I think there's no point in supporting the `Coord = i32` in this multi-touch handling.

File: internal/core/window.rs [hunk 10]
```diff
@@ -438,13 +438,1023 @@ struct WindowPinnedFields {
...
+            (a.position.x + b.position.x) / (2 as crate::Coord),
+            (a.position.y + b.position.y) / (2 as crate::Coord),
+        ))
+    }
+
+    /// Returns (distance, angle_in_degrees) between the two gesture fingers.
+    fn gesture_geometry(&self) -> Option<(f32, f32)> {
```
Comment (tronical) [inline internal/core/window.rs]: Similarly here, I think you can use Euclid's subtraction to compute the distance vector and then use `.length()` to obtain the length. For the angle, I think that you can use `angle_from_x_axis()` to obtain the angle as `Angle`, which I would suggest to return. On that one you can call for example `.to_degress()` if you need the angle in degrees.

File: internal/core/window.rs [hunk 11]
```diff
@@ -438,13 +438,1023 @@ struct WindowPinnedFields {
...
+    fn snapshot_initial_geometry(&mut self) {
+        if let Some((dist, angle)) = self.gesture_geometry() {
+            self.initial_distance = dist;
+            self.last_scale = 1.0;
+            self.last_angle = angle;
+        }
+    }
+
+    /// Normalize an angle difference to [-180, 180].
+    fn normalize_angle(degrees: f32) -> f32 {
```
Comment (tronical) [inline internal/core/window.rs]: I think if you're using Euclid's `Angle` type instead, then this function isn't needed and you can use `.signed()` to obtain a `[-pi..pi]` range and `.to_degrees()` if needed. Less custom code, more re-use from Euclid :)

File: internal/core/window.rs [hunk 12]
```diff
@@ -438,13 +438,1023 @@ struct WindowPinnedFields {
...
+        &mut self,
+        id: u64,
+        position: LogicalPoint,
+        click_interval: core::time::Duration,
+        events: &mut TouchEventBuffer,
+    ) {
+        let is_double_tap = if self.active_touches.len() == 0 {
+            // Check for double-tap before inserting, so a second finger
+            // arriving during the double-tap touch doesn't see a stale entry.
+            if let (Some(last_time), Some(last_pos)) = (self.last_tap_time, self.last_tap_position)
```
Comment (tronical) [inline internal/core/window.rs]: If the time stamp was part of the `TouchPoint`, then `last_tap` could be a `TouchPoint` instead of just a position, and you'd have only one field. Less state to manage as well, as I think you effectively assert that last_tap_time and last_tap_position are either both Some or both None, but if the type system can assert that I think the code will be more robust.

File: internal/core/window.rs [hunk 13]
```diff
@@ -438,13 +438,1023 @@ struct WindowPinnedFields {
...
+            {
+                let now = crate::animations::Instant::now();
```
Comment (tronical) [inline internal/core/window.rs:671]: If the `process` functions would also take a `TouchPoint` that would have the time stamp, then you wouldn't have to call `Instance::now()` in this place and the signatures of your process functions would become shorter.

**Review by tronical — COMMENTED** [review]
I tested this on an i.mx95 with the linuxkms backend and it works - yay :). I did a first pass through the code and I think there are quite a few things that could be simplified: Less custom vector operations and more use of Euclid, as well as combining more fields together in structs or enumeration variants that belong together - to make the state machine more robust and reduce the amount of function parameters.

**Comment by tilladam** [comment]
Thanks for the thorough review @tronical! I've addressed all the feedback in individual commits:

1. **Android: use PhysicalPosition and to_logical** — restored the `PhysicalPosition` → `to_logical()` pattern
2. **LinuxKMS: document why last_touch_positions is needed** — added comment explaining touch-up events lack coordinates
3. **LinuxKMS: move touch position helpers into LibInputHandler** — `set_touch_pos`/`take_touch_pos` are now methods on `&mut self`
4. **Make TouchEventBuffer Clone, remove test-only iter()** — tests now use `clone().into_iter()` instead of special `iter()`
5. **Add id field to TouchPoint, simplify TouchMap API** — `TouchPoint { id, position }` replaces `(u64, TouchPoint)` tuples; `first()` returns `Option<&TouchPoint>`
6. **Combine last_tap_time/last_tap_position into single last_tap** — single `Option<(Instant, LogicalPoint)>` enforced by the type system
7. **Use Euclid for midpoint, distance, and angle** — `lerp` for midpoint, `Vector2D::length()` for distance, `angle_from_x_axis()` + `Angle::signed()` for rotation; removed custom `normalize_angle`
8. **Move gesture fields into GestureRecognitionState enum variants** — `finger_ids`, `initial_distance`, `last_scale`, `last_angle` now live inside `TwoFingersDown`/`Pinching` variants

File: internal/backends/linuxkms/calloop_backend/input.rs (hunk 3 above)
Comment (tronical) [reply internal/backends/linuxkms/calloop_backend/input.rs:162]: Ah, I See that doesn't work, because we already mutably borrow some fields of self :(

**Review by tronical — APPROVED** [review]
